File: ts/test/entity/daily_post/DailyPostEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { NavalnyarchiveSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('DailyPostEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when NAVALNYARCHIVE_TEST_LIVE=TRUE.
  afterEach(liveDelay('NAVALNYARCHIVE_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = NavalnyarchiveSDK.test()
    const ent = testsdk.DailyPost()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.NAVALNYARCHIVE_TEST_LIVE
    for (const op of ['list']) {
      if (maybeSkipControl(t, 'entityOp', 'daily_post.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set NAVALNYARCHIVE_TEST_DAILY_POST_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let daily_post_ref01_data = Object.values(setup.data.existing.daily_post)[0] as any

    // LIST
    const daily_post_ref01_ent = client.DailyPost()
    const daily_post_ref01_match: any = {}

    const daily_post_ref01_list = await daily_post_ref01_ent.list(daily_post_ref01_match)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/daily_post/DailyPostTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = NavalnyarchiveSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['daily_post01','daily_post02','daily_post03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['NAVALNYARCHIVE_TEST_DAILY_POST_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'NAVALNYARCHIVE_TEST_DAILY_POST_ENTID': idmap,
    'NAVALNYARCHIVE_TEST_LIVE': 'FALSE',
    'NAVALNYARCHIVE_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['NAVALNYARCHIVE_TEST_DAILY_POST_ENTID']

  const live = 'TRUE' === env.NAVALNYARCHIVE_TEST_LIVE

  if (live) {
    client = new NavalnyarchiveSDK(merge([
      {
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.NAVALNYARCHIVE_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  

File: rb/test/daily_post_entity_test.rb
# DailyPost entity test

require "minitest/autorun"
require "json"
require_relative "../Navalnyarchive_sdk"
require_relative "runner"

class DailyPostEntityTest < Minitest::Test
  def test_create_instance
    testsdk = NavalnyarchiveSDK.test(nil, nil)
    ent = testsdk.DailyPost(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "daily_post" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = NavalnyarchiveSDK.test(seed, nil)
    seen = base.DailyPost(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = NavalnyarchiveConfig.make_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = NavalnyarchiveSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.DailyPost(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = daily_post_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "daily_post." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set NAVALNYARCHIVE_TEST_DAILY_POST_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    daily_post_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.daily_post")))
    daily_post_ref01_data = nil
    if daily_post_ref01_data_raw.length > 0
      daily_post_ref01_data = Helpers.to_map(daily_post_ref01_data_raw[0][1])
    end

    # LIST
    daily_post_ref01_ent = client.DailyPost(nil)
    daily_post_ref01_match = {}

    daily_post_ref01_list_result = daily_post_ref01_ent.list(daily_post_ref01_match, nil)
    assert daily_post_ref01_list_result.is_a?(Array)

  end
end

def daily_post_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "daily_post", "DailyPostTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = NavalnyarchiveSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["daily_post01", "daily_post02", "daily_post03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["NAVALNYARCHIVE_TEST_DAILY_POST_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "NAVALNYARCHIVE_TEST_DAILY_POST_ENTID" => idmap,
    "NAVALNYARCHIVE_TEST_LIVE" => "FALSE",
    "NAVALNYARCHIVE_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["NAVALNYARCHIVE_TEST_DAILY_POST_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["NAVALNYARCHIVE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = NavalnyarchiveSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["NAVALNYARCHIVE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["NAVALNYARCHIVE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
